File: coldatomlab/web/precision.js
"use strict";
(() => {
  const arms=["ideal","finite"],cases=["baseline","more_atoms","more_shots"];
  const convention="v0.14 readout; independent full-scan repetitions; LCG blocks: trial,case,setting,arm; circular statistics conditional on resolved fits";
  const defaults={readout:{...Readout.defaults,base:{...Readout.defaults.base,atoms:20},shots:64},more_atoms:80,more_shots:256,repetitions:100};
  function validate(input){
    if(Object.keys(input).sort().join()!==Object.keys(defaults).sort().join())throw new Error("Unknown precision setting.");
    const p={...input,readout:Readout.validate(input.readout)};
    for(const [k,lo,hi] of [["more_atoms",p.readout.base.atoms+1,100],["more_shots",p.readout.shots+1,4096],["repetitions",20,300]])
      if(!Number.isInteger(p[k])||p[k]<lo||p[k]>hi)throw new Error(`${k} must be an integer between ${lo} and ${hi}.`);
    if(drawsPerTrial(p)*p.repetitions>16000000)throw new Error("Workload exceeds 16 million count draws. Reduce repetitions, settings or shots.");
    return p;
  }
  const drawsPerTrial=p=>2*p.readout.points*(2*p.readout.shots+p.more_shots);
  const wrap=x=>Math.atan2(Math.sin(x),Math.cos(x));
  // Affine LCG skip-ahead, preserving unsigned 32-bit arithmetic exactly.
  function jump(seed,steps){let a=1664525,c=1013904223,x=seed>>>0;
    while(steps>0){if(steps%2===1)x=(Math.imul(a,x)+c)>>>0;c=(Math.imul(a,c)+c)>>>0;a=Math.imul(a,a)>>>0;steps=Math.floor(steps/2);}return x;}
  function configs(p){return cases.map(name=>({name,plan:Readout.validate({...p.readout,base:{...p.readout.base,atoms:name==="more_atoms"?p.more_atoms:p.readout.base.atoms},shots:name==="more_shots"?p.more_shots:p.readout.shots})}));}
  function modelCase(item){const {name,plan}=item,source=Readout.incoming(plan);
    const points=Array.from({length:plan.points},(_,i)=>{const r=Readout.states(plan,source,i);return {index:i,alpha:r.alpha,arms:Object.fromEntries(arms.map(a=>[a,r.arms[a]]))};});
    const model_fits=Object.fromEntries(arms.map(a=>[a,Readout.fit(points.map(r=>({alpha:r.alpha,z:2*r.arms[a].mean_left/plan.base.atoms-1,variance_z_mean:0})))]));
    const eligible=plan.base.initial==="coherent"&&plan.base.left_fraction===.5&&plan.base.interaction_hz===0;
    return {name,plan,source,points,model_fits,ideal_scatter_guide:eligible?Math.sqrt(3/(2*plan.base.atoms*plan.points*plan.shots)):null};
  }
  function trial(p,models,index){if(!Number.isInteger(index)||index<0||index>=p.repetitions)throw new Error("Invalid repetition index.");
    let seed=jump(p.readout.seed,index*drawsPerTrial(p));const result={index,cases:{}};
    for(const model of models){const S=model.plan.shots,N=model.plan.base.atoms,values=Object.fromEntries(arms.map(a=>[a,{measurements:[],fit:null}]));
      for(const r of model.points)for(const a of arms){const sample=TwoMode.sample(r.arms[a],S,seed);values[a].measurements.push({seed,counts:sample.counts});seed=jump(seed,S);}
      for(const a of arms)values[a].fit=Readout.fit(values[a].measurements.map((m,i)=>({alpha:model.points[i].alpha,...Readout.statistics(m.counts,N)})));
      result.cases[model.name]=values;
    }return result;
  }
  function summary(model,records,arm){const fits=records.map(r=>r.cases[model.name][arm].fit),valid=fits.filter(f=>f.phase!==null),n=valid.length,truth=model.source.phase;
    const result={attempted:fits.length,resolved:n,unresolved:fits.length-n,circular_bias:null,circular_scatter:null,wrapped_rmse:null,rms_reported_se:null,scatter_over_se:null,within_one_se:null,model_bias:truth!==null&&model.model_fits[arm].phase!==null?wrap(model.model_fits[arm].phase-truth):null};
    if(n<2)return result;
    const re=valid.reduce((s,f)=>s+Math.cos(f.phase),0)/n,im=valid.reduce((s,f)=>s+Math.sin(f.phase),0)/n,R=Math.min(1,Math.hypot(re,im));
    if(R>1e-8){result.circular_scatter=Math.sqrt(Math.max(0,-2*Math.log(R)));if(truth!==null)result.circular_bias=wrap(Math.atan2(im,re)-truth);}
    result.rms_reported_se=Math.sqrt(valid.reduce((s,f)=>s+f.phase_se**2,0)/n);
    if(result.rms_reported_se>0&&result.circular_scatter!==null)result.scatter_over_se=result.circular_scatter/result.rms_reported_se;
    if(truth!==null){result.wrapped_rmse=Math.sqrt(valid.reduce((s,f)=>s+wrap(f.phase-truth)**2,0)/n);result.within_one_se=valid.filter(f=>Math.abs(wrap(f.phase-truth))<=f.phase_se).length/n;}
    return result;
  }
  function report(p,models,records,status){return {schema:"coldatomlab-precision-v1",version:"0.15.0",convention,plan:p,status,models,records,summary:Object.fromEntries(models.map(m=>[m.name,Object.fromEntries(arms.map(a=>[a,summary(m,records,a)]))]))};}
  globalThis.Precision={arms,cases,convention,defaults,validate,drawsPerTrial,wrap,jump,configs,modelCase,trial,summary,report};
})();

File: coldatomlab/precision.py
"""Independent repeated-readout reference and strict reproducibility verification."""

import math

import numpy as np

from . import readout
from .echo import compare_tree
from .twomode import sample

ARMS = ("ideal", "finite")
CASES = ("baseline", "more_atoms", "more_shots")
CONVENTION = (
    "v0.14 readout; independent full-scan repetitions; LCG blocks: trial,case,setting,arm; "
    "circular statistics conditional on resolved fits"
)


def draws_per_trial(p):
    return 2 * p["readout"]["points"] * (2 * p["readout"]["shots"] + p["more_shots"])


def validate(plan):
    if set(plan) != {"readout", "more_atoms", "more_shots", "repetitions"}:
        raise ValueError("Unknown precision settings.")
    p = plan | dict(readout=readout.validate(plan["readout"]))
    for k, lo, hi in [
        ("more_atoms", p["readout"]["base"]["atoms"] + 1, 100),
        ("more_shots", p["readout"]["shots"] + 1, 4096),
        ("repetitions", 20, 300),
    ]:
        x = p[k]
        if type(x) not in (int, float) or not math.isfinite(x) or int(x) != x or not lo <= x <= hi:
            raise ValueError(f"Invalid {k}.")
        p[k] = int(x)
    if draws_per_trial(p) * p["repetitions"] > 16000000:
        raise ValueError("Workload exceeds 16 million count draws.")
    return p


def jump(seed, steps):
    # Compose integer affine maps, independently of the browser Math.imul operations.
    a, c, x = 1664525, 1013904223, seed
    while steps:
        if steps & 1:
            x = (a * x + c) % 2**32
        a, c = a * a % 2**32, (a * c + c) % 2**32
        steps //= 2
    return x


def configs(p):
    return [
        dict(
            name=name,
            plan=readout.validate(
                p["readout"]
                | dict(
                    base=p["readout"]["base"]
                    | dict(
                        atoms=p["more_atoms"]
                        if name == "more_atoms"
                        else p["readout"]["base"]["atoms"]
                    ),
                    shots=p["more_shots"] if name == "more_shots" else p["readout"]["shots"],
                )
            ),
        )
        for name in CASES
    ]


def model_case(item):
    name, plan = item["name"], item["plan"]
    source = readout.incoming(plan)
    points = []
    for i in range(plan["points"]):
        r = readout.states(plan, source, i)
        points.append(dict(index=i, alpha=r["alpha"], arms={a: r["arms"][a] for a in ARMS}))
    model_fits = {
        a: readout.fit(
            [
                dict(
                    alpha=r["alpha"],
                    z=2 * r["arms"][a]["mean_left"] / plan["base"]["atoms"] - 1,
                    variance_z_mean=0,
                )
                for r in points
            ]
        )
        for a in ARMS
    }
    b = plan["base"]
    eligible = b["initial"] == "coherent" and b["left_fraction"] == 0.5 and b["interaction_hz"] == 0
    return dict(
        name=name,
        plan=plan,
        source=source,
        points=points,
        model_fits=model_fits,
        ideal_scatter_guide=math.sqrt(3 / (2 * b["atoms"] * plan["points"] * plan["shots"]))
        if eligible
        else None,
    )


def trial(p, models, index):
    seed = jump(p["readout"]["seed"], index * draws_per_trial(p))
    result = dict(index=index, cases={})
    for model in models:
        shots, n = model["plan"]["shots"], model["plan"]["base"]["atoms"]
        values = {a: dict(measurements=[], fit=None) for a in ARMS}
        for r in model["points"]:
            for a in ARMS:
                counts = sample(r["arms"][a], shots, seed)["counts"]
                values[a]["measurements"].append(dict(seed=seed, counts=counts))
                seed = jump(seed, shots)
        for a in ARMS:
            values[a]["fit"] = readout.fit(
                [
                    dict(alpha=model["points"][i]["alpha"], **readout.statistics(m["counts"], n))
                    for i, m in enumerate(values[a]["measurements"])
                ]
            )
        result["cases"][model["name"]] = values
    return result


def wrap(x):
    return float(np.angle(np.exp(1j * x)))


def summary(model, records, arm):
    fits = [r["cases"][model["name"]][arm]["fit"] for r in records]
    valid = [f for f in fits if f["phase"] is not None]
    n, truth = len(valid), model["source"]["phase"]
    model_phase = model["model_fits"][arm]["phase"]
    result = dict(
        attempted=len(fits),
        resolved=n,
        unresolved=len(fits) - n,
        circular_bias=None,
        circular_scatter=None,
        wrapped_rmse=None,
        rms_reported_se=None,
        scatter_over_se=None,
        within_one_se=None,
        model_bias=wrap(model_phase - truth)
        if truth is not None and model_phase is not None
        else None,
    )
    if n < 2:
        return result
    phases = np.array([f["phase"] for f in valid])
    ses = np.array([f["phase_se"] for f in valid])
    mean = np.exp(1j * phases).mean()
    length = min(1, abs(mean))
    if length > 1e-8:
        result["circular_scatter"] = math.sqrt(max(0, -2 * math.log(length)))
        if truth is not None:
            result["circular_bias"] = wrap(float(np.angle(mean)) - truth)
    result["rms_reported_se"] = float(np.sqrt(np.mean(ses**2)))
    if result["rms_reported_se"] > 0 and result["circular_scatter"] is not None:
        result["scatter_over_se"] = result["circular_scatter"] / result["rms_reported_se"]
    if truth is not None:
        errors = np.angle(np.exp(1j * (phases - truth)))
        result["wrapped_rmse"] = float(np.sqrt(np.mean(errors**2)))
        result["within_one_se"] = float(np.mean(abs(errors) <= ses))
    return result


def report(p, models, records, status):
    return dict(
        schema="coldatomlab-precision-v1",
        version="0.15.0",
        convention=CONVENTION,
        plan=p,
        status=status,
        models=models,
        records=records,
        summary={m["name"]: {a: summary(m, records, a) for a in ARMS} for m in models},
    )


def verify_export(data):
    p = validate(data["plan"])
    n = len(data["records"])
    if (
        data["status"] not in ("complete", "cancelled")
        or not 1 <= n <= p["repetitions"]
        or (data["status"] == "complete" and n != p["repetitions"])
    ):
        raise ValueError("Invalid completed repetition prefix.")
    models = [model_case(item) for item in configs(p)]
    records = [trial(p, models, i) for i in range(n)]
    expected = report(p, models, records, data["status"])
    compare_tree(data, expected, "Precision")
    errors, drift = [], []
    for saved, reference in zip(data["models"], models, strict=True):
        ss = [saved["source"]] + [r["arms"][a] for r in saved["points"] for a in ARMS]
        rs = [reference["source"]] + [r["arms"][a] for r in reference["points"] for a in ARMS]
        for x, y in zip(ss, rs, strict=True):
            psi = np.array(x["real"]) + 1j * np.array(x["imag"])
            oracle = np.array(y["real"]) + 1j * np.array(y["imag"])
            errors.append(float(np.linalg.norm(psi - oracle)))
            drift.append(float(abs(np.vdot(psi, psi).real - 1)))
    if max(errors) > 5e-8 or max(drift) > 1e-9:
        raise ValueError("Precision model state or norm tolerance exceeded.")
    return dict(
        verified=True,
        repetitions_verified=n,
        cases_verified=3,
        status=data["status"],
        maximum_complex_state_l2=max(errors),
        maximum_norm_drift=max(drift),
        reference="Independent NumPy propagation, new count draws, fits and circular statistics",
    )
